### PR TITLE
Support ddp_fork strategy with native AMP by attempting NVML-based CUDA availability assessment

### LINKED ISSUE #14981: Support ddp_fork strategy with native AMP by attempting NVML-based CUDA availability assessment
## 🚀 Feature

`ddp_fork` (and associated alias strategies) cannot currently be used along with native AMP due to the invocation of the CUDA Runtime API within the call to `GradScaler` in the `NativeMixedPrecisionPlugin`:

https://github.com/Lightning-AI/lightning/blob/c059db446e7bfea03fba91e598ad503f0d1c6581/src/pytorch_lightning/plugins/precision/native_amp.py#L53

which in turn initializes CUDA and poisons subsequent forks.

It may be possible with a future version of PyTorch to alter the default behavior of `torch.cuda.is_available()` to use an NVML-based CUDA assessment throughout Lightning. In the meantime, patching  `torch.cuda.is_available()` with Lightning's implementation of the upstream NVML-based assessment can unlock this functionality.

I'll be opening a PR shortly that patches `torch.cuda.is_available()` within `NativeMixedPrecisionPlugin` (both Lite and PL versions) and adds a standalone test for the `ddp_fork` strategy in a CUDA and AMP context (adding a standalone test only for PL given how expensive the standalone multi-gpu tests can be).

### Motivation

Many users will use AMP within the context of jupyter notebooks, where if using multiple GPUS,  `ddp_fork` will be important to support.

### Pitch

Allow the use of AMP within the context of jupyter notebooks, where if using multiple GPUS,  `ddp_fork` will be important to support.
I will open a small PR shortly that makes this available.

### Additional context
There's a related [PR](https://github.com/pytorch/pytorch/pull/85951) in PyTorch currently that may allow the requested modification of `torch.cuda.is_available()` throughout Lightning without needing to patch the function or add Lightning's own NVML-based assessment (once the relevant version of PyTorch is the minimum)

### PR DESCRIPTION
**Pull request by speediedan**
## What does this PR do?
Fixes #14981 

`ddp_fork` (and associated alias strategies) cannot currently be used along with native AMP due to the invocation of the CUDA Runtime API within the call to `GradScaler` in the `NativeMixedPrecisionPlugin`:

https://github.com/Lightning-AI/lightning/blob/c059db446e7bfea03fba91e598ad503f0d1c6581/src/pytorch_lightning/plugins/precision/native_amp.py#L53

which in turn initializes CUDA and poisons subsequent forks.

It may be possible with a future version of PyTorch to alter the default behavior of `torch.cuda.is_available()` to use an NVML-based CUDA assessment throughout Lightning. In the meantime, patching  `torch.cuda.is_available()` with Lightning's implementation of the upstream NVML-based assessment can unlock this functionality.

This PR patches `torch.cuda.is_available()` within `NativeMixedPrecisionPlugin` (both Lite and PL versions) and adds a standalone test for the `ddp_fork` strategy in a CUDA and native AMP context (adding a standalone test only for PL given how expensive the standalone multi-gpu tests can be).

The use of `ddp_fork` (and its associated alias strategies) with native AMP in context (e.g. in the context of  Jupyter notebooks) should be possible after this PR. Note that this PR does not address or pertain to bf16 based AMP support.

### Additional context
There's a related [PR](https://github.com/pytorch/pytorch/pull/85951) in PyTorch currently that may allow the requested modification of `torch.cuda.is_available()` throughout Lightning without needing to patch the function or add Lightning's own NVML-based assessment (once the relevant version of PyTorch is the minimum)

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tests_pytorch/helpers/runif.py [hunk 1]
```diff
@@ -249,6 +251,17 @@ def __new__(
             conditions.append(not _HIVEMIND_AVAILABLE or sys.platform in ("win32", "darwin"))
             reasons.append("Hivemind")
 
+        if min_cuda_gpus_no_init:
```
Comment (carmocca) [inline tests/tests_pytorch/helpers/runif.py:254]: We can replace the logic in the block above with his. No need for different flags.


Also, can you apply this change to the Lite `RunIf`?

File: tests/tests_pytorch/models/test_ddp_fork_amp.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import tests_pytorch.helpers.utils as tutils
+from pytorch_lightning import Trainer
+from pytorch_lightning.demos.boring_classes import RandomDataset
+from tests_pytorch.helpers.runif import RunIf
+from tests_pytorch.helpers.test_models import AMPTestModel
+
+
+# needs to be standalone to avoid other processes initializing CUDA
+@RunIf(min_cuda_gpus_no_init=2, skip_windows=True, min_torch="1.12", standalone=True)
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_ddp_fork_amp.py:24]: Why is `min_torch` set?

File: tests/tests_pytorch/models/test_ddp_fork_amp.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        accelerator="gpu",
+        devices=2,
+        strategy="ddp_fork",
+        precision=16,
+    )
+
+    model = AMPTestModel()
+    trainer.fit(model)
+    trainer.test(model)
+    trainer.predict(model, DataLoader(RandomDataset(32, 64)))
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_ddp_fork_amp.py:41]: Instead of this broad test, can you write one that does

```
plugin = NativeMixedPrecisionPlugin()
_check_in_bad_fork()  # would raise an error if not patched
```

File: src/lightning_lite/accelerators/cuda.py [hunk 4]
```diff
@@ -77,6 +78,25 @@ def _get_all_available_cuda_gpus() -> List[int]:
     return list(range(num_cuda_devices()))
 
 
+@contextmanager
+def _patch_cuda_is_available() -> Generator:
+    """Context manager that safely patches :func:`torch.cuda.is_available` with its NVML-based version if
+    possible."""
+    orig_check = None
+    new_check = torch.cuda.device_count if _TORCH_GREATER_EQUAL_1_13 else _device_count_nvml
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:86]: Can you make it just exit if >=1.13? No patching at all

File: tests/tests_pytorch/helpers/runif.py (hunk 1 above)
Comment (speediedan) [reply tests/tests_pytorch/helpers/runif.py]: Was wondering if it'd be okay to change the behavior for all tests, SGTM! I've made both suggested changes in 
a30eca4

File: tests/tests_pytorch/models/test_ddp_fork_amp.py (hunk 2 above)
Comment (speediedan) [reply tests/tests_pytorch/models/test_ddp_fork_amp.py]: Good catch! Removed.
btw, this test won't run with torch == 1.12.0 because of the issue fixed [here](https://github.com/pytorch/pytorch/pull/80899/files#diff-dd8250f2740baa59fd4ba7431ad9d43f560d801ae2226ad492c2be9858d0b660) in 1.12.1 but I don't think that's an issue

File: tests/tests_pytorch/models/test_ddp_fork_amp.py (hunk 3 above)
Comment (speediedan) [reply tests/tests_pytorch/models/test_ddp_fork_amp.py]: Done, added this test version in a30eca4.

File: src/lightning_lite/accelerators/cuda.py (hunk 4 above)
Comment (speediedan) [reply src/lightning_lite/accelerators/cuda.py:86]: In PyTorch 1.13 as it stands, `torch.cuda.is_available()` still uses a CUDA Runtime API based availability assessment and therefore poisons subsequent forks. I have a [PR open](https://github.com/pytorch/pytorch/pull/85951) (that @awaelchli has been awesome reviewing, thanks man!) in PyTorch currently that may allow us to request NVML-based behavior for `torch.cuda.is_available()` throughout Lightning without needing to patch the function or add Lightning's own NVML-based assessment implementation. I don't know yet if that behavioral enhancement will make it into 1.13 though so I think we need to assume continued patching with 1.13 for now.

Comment (awaelchli) [reply src/lightning_lite/accelerators/cuda.py:86]: Exactly! I think we need to first know if it really goes in to 1.13. We can open an issue on Lighting as a reminder that we want to evaluate this again when 1.13 is out.

File: tests/tests_pytorch/models/test_amp.py [hunk 5]
```diff
@@ -24,47 +23,7 @@
 from pytorch_lightning import Trainer
 from pytorch_lightning.demos.boring_classes import BoringModel, RandomDataset
 from tests_pytorch.helpers.runif import RunIf
-
-
-class AMPTestModel(BoringModel):
```
Comment (awaelchli) [inline tests/tests_pytorch/models/test_amp.py:29]: let's put it back if we don't need to move it? I think the other test you wrote is perfectly fine :)

**Review by awaelchli — APPROVED** [review]

File: src/lightning_lite/accelerators/cuda.py [hunk 6]
```diff
@@ -77,6 +78,25 @@ def _get_all_available_cuda_gpus() -> List[int]:
...
+    if hasattr(torch._C, "_cuda_getDeviceCount") and _device_count_nvml() >= 0:
+        # we can safely patch is_available if both torch has CUDA compiled and the NVML count is succeeding
+        # otherwise, patching is_available could lead to attribute errors or infinite recursion
+        orig_check = torch.cuda.is_available
+        torch.cuda.is_available = new_check  # type: ignore[assignment]
+    try:
+        yield
+    finally:
+        if orig_check:
+            torch.cuda.is_available = orig_check
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:97]: Can't we do this? Where we avoid checking None

```suggestion
        try:
            yield
        finally:
            torch.cuda.is_available = orig_check
```

File: src/lightning_lite/accelerators/cuda.py [hunk 7]
```diff
@@ -77,6 +78,25 @@ def _get_all_available_cuda_gpus() -> List[int]:
     return list(range(num_cuda_devices()))
 
 
+@contextmanager
+def _patch_cuda_is_available() -> Generator:
+    """Context manager that safely patches :func:`torch.cuda.is_available` with its NVML-based version if
+    possible."""
+    orig_check = None
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:85]:
```suggestion
```

File: src/lightning_lite/accelerators/cuda.py (hunk 4 above)
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:86]: Can you move the definition inside the `if` that uses it?

File: src/lightning_lite/accelerators/cuda.py [hunk 8]
```diff
@@ -77,6 +78,25 @@ def _get_all_available_cuda_gpus() -> List[int]:
...
+    new_check = torch.cuda.device_count if _TORCH_GREATER_EQUAL_1_13 else _device_count_nvml
+
+    if hasattr(torch._C, "_cuda_getDeviceCount") and _device_count_nvml() >= 0:
+        # we can safely patch is_available if both torch has CUDA compiled and the NVML count is succeeding
+        # otherwise, patching is_available could lead to attribute errors or infinite recursion
+        orig_check = torch.cuda.is_available
+        torch.cuda.is_available = new_check  # type: ignore[assignment]
```
Comment (carmocca) [inline src/lightning_lite/accelerators/cuda.py:92]: I'm a bit confused. Why do we want to assign `device_count()` to `is_available`? Wouldn't we need to convert to bool?

File: tests/tests_lite/helpers/runif.py [hunk 9]
```diff
@@ -71,9 +72,10 @@ def __new__(
         """
         conditions = []
         reasons = []
-
         if min_cuda_gpus:
-            conditions.append(torch.cuda.device_count() < min_cuda_gpus)
+            # defer CUDA initialization if possible, supporting tests that will attempt forks after NVML-based CUDA
+            # availability checks
+            conditions.append(num_cuda_devices() < min_cuda_gpus)
```
[the remark this answers is not in the document]
Comment (speediedan) [reply tests/tests_lite/helpers/runif.py]: makes sense, I've removed the overzealous documentation :smile:

File: tests/tests_pytorch/models/test_amp.py (hunk 5 above)
Comment (speediedan) [reply tests/tests_pytorch/models/test_amp.py]: Makes sense, moved it back.

File: src/lightning_lite/accelerators/cuda.py (hunk 4 above)
Comment (speediedan) [reply src/lightning_lite/accelerators/cuda.py]: Good suggestion, done!

File: src/lightning_lite/accelerators/cuda.py (hunk 8 above)
Comment (speediedan) [reply src/lightning_lite/accelerators/cuda.py]: Since the returned `int` would never be negative in this context, I was avoiding the unnecessary check using the truth value of the returned int. Using `is_cuda_available` may be cleaner though so using that with the latest commit (after rebasing).

File: tests/tests_pytorch/models/test_ddp_fork_amp.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+# needs to be standalone to avoid other processes initializing CUDA
+@RunIf(min_cuda_gpus=2, skip_windows=True, standalone=True)
+def test_amp_gpus_ddp_fork():
+    """Ensure the use of native AMP with `ddp_fork` (or associated alias strategies) does not generate CUDA
+    initialization errors."""
+    _ = NativeMixedPrecisionPlugin(precision=16, device="cuda")
+    with multiprocessing.get_context("fork").Pool(1) as pool:
+        in_bad_fork = pool.apply(torch.cuda._is_in_bad_fork)
+    assert not in_bad_fork
```
Comment (carmocca) [inline tests/tests_pytorch/models/test_ddp_fork_amp.py:30]: Would this be enough? Then we wouldn't need 2 GPUs in the test
```suggestion
    assert not torch.cuda.is_initialized()
```

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: tests/tests_pytorch/models/test_ddp_fork_amp.py (hunk 10 above)
Comment (awaelchli) [reply tests/tests_pytorch/models/test_ddp_fork_amp.py:30]: @speediedan Yes, I think you are right. For this reason my check here is probably also too weak: https://github.com/Lightning-AI/lightning/blob/7fed7a12c56c6b0e08f0223f68d7d311ea7ac90f/src/lightning_lite/strategies/launchers/multiprocessing.py#L178
